### PR TITLE
add retry flag for notify providers

### LINKED ISSUE #311: using an unknown template makes AM flood your $alerttarget
##### Problem:
When specifying an unknown template at least to the "html"-field of a receiver makes AM flood the receiver's inbox with mails where the according field (in case of " html:" the mail body) is empty.

##### Steps to reproduce:
use e.g. `html: '{{ email.shinanigans.html }}'` in a receiver-block and provoke an alert for that receiver.

@fabxc

### PR DESCRIPTION
**Pull request by brancz**
The retry flag allows an integration to specify whether a retry can
potentially be solved or if the error is likely not going to recover.
For example invalid authentication is likely a wrong configuration and
therefore a retry would not make sense, while a server error is likely
a temporary problem and can potentially be solved on the next retry.

It would be good to review all the added flags that enable/disable
retry, especially the ones in the email notifier.

@fabxc @beorn7 @brian-brazil 

Fixes #311

### CODE REVIEW

File: notify/impl.go [hunk 1]
```diff
@@ -153,20 +155,24 @@ func (w *Webhook) Notify(ctx context.Context, alerts ...*types.Alert) error {
...
 	var buf bytes.Buffer
 	if err := json.NewEncoder(&buf).Encode(msg); err != nil {
-		return err
+		return false, err
 	}
 
 	resp, err := ctxhttp.Post(ctx, http.DefaultClient, w.URL, contentTypeJSON, &buf)
 	if err != nil {
-		return err
+		return false, err
```
Comment (brancz) [inline notify/impl.go]: wdyt about these cases, recoverable or not?

Comment (fabxc) [inline notify/impl.go]: I went through the code chain again. This is mostly not recoverable (malformed requests etc.) – however, there are some points like the server returning an HTTP response to an HTTPS request etc., that could in theory be resolved on the side of the integration.

I'd vote to make it retry until someone reports a problem. Inspecting the error for further differentiation is probably not worth the effort. Retrying is not exactly costly for us.

File: notify/impl.go [hunk 2]
```diff
@@ -153,20 +155,24 @@ func (w *Webhook) Notify(ctx context.Context, alerts ...*types.Alert) error {
...
 	}
 	resp.Body.Close()
 
-	if resp.StatusCode/100 != 2 {
-		return fmt.Errorf("unexpected status code %v from %s", resp.StatusCode, w.URL)
+	if resp.StatusCode/100 == 5 {
+		return true, fmt.Errorf("unexpected status code %v from %s", resp.StatusCode, w.URL)
+	}
+
+	if resp.StatusCode/100 != 2 && resp.StatusCode/100 != 3 {
```
Comment (fabxc) [inline notify/impl.go]: Redirects are resolved by the HTTP client, so we cannot encounter them at this point anymore. Also, if we had received a redirect at this point, we would not have sent anything and returns `true, nil` would not really be sane.

File: notify/impl.go [hunk 3]
```diff
@@ -390,24 +399,31 @@ func (n *PagerDuty) Notify(ctx context.Context, as ...*types.Alert) error {
...
 	resp, err := ctxhttp.Post(ctx, http.DefaultClient, n.conf.URL, contentTypeJSON, &buf)
 	if err != nil {
-		return err
+		return false, err
 	}
 	resp.Body.Close()
 
+	// retry can solve the issue on 403 (rate limiting) and 5xx response codes
+	// https://v2.developer.pagerduty.com/docs/trigger-events
+	if resp.StatusCode == 403 || resp.StatusCode/100 == 5 {
```
Comment (fabxc) [inline notify/impl.go]: Oh my god – 403 for rate limiting?
I was just thinking about generalizing the retry inspection in its own function...

File: notify/impl.go [hunk 4]
```diff
@@ -651,40 +680,45 @@ func (n *OpsGenie) Notify(ctx context.Context, as ...*types.Alert) error {
...
+		return false, fmt.Errorf("unexpected status code %v", resp.StatusCode)
 	} else if resp.StatusCode/100 != 2 {
 		body, _ := ioutil.ReadAll(resp.Body)
 		log.With("incident", key).Debugf("unexpected OpsGenie response from %s (POSTed %s), %s: %s",
 			apiURL, msg, resp.Status, body)
-		return fmt.Errorf("unexpected status code %v", resp.StatusCode)
+		return false, fmt.Errorf("unexpected status code %v", resp.StatusCode)
 	}
-	return nil
+	return true, nil
```
Comment (fabxc) [inline notify/impl.go]: Mh, shouldn't this default to false everywhere when everything went alright?

File: notify/impl.go (hunk 3 above)
Comment (brancz) [inline notify/impl.go]: pd is even one of the better APIs if you ask me ... I guess a function per provider to decide whether to retry or not makes sense, but might not work for all providers..

File: notify/impl.go (hunk 2 above)
Comment (brancz) [inline notify/impl.go]: yes, of course .. missed to remove that one, it was one of my first thoughts and removed it in all the other ones..

File: notify/notify.go [hunk 5]
```diff
@@ -575,9 +575,13 @@ func (r RetryStage) Exec(ctx context.Context, alerts ...*types.Alert) (context.C
...
 		select {
 		case <-tick.C:
-			if err := r.integration.Notify(ctx, alerts...); err != nil {
+			if retry, err := r.integration.Notify(ctx, alerts...); err != nil {
 				numFailedNotifications.WithLabelValues(r.integration.name).Inc()
 				log.Warnf("Notify attempt %d failed: %s", i, err)
+				if !retry {
+					log.Warnf("Cancelling notify retry due to unrecoverable error: %s", err)
+					return ctx, alerts, nil
+				}
```
Comment (fabxc) [inline notify/notify.go]: We should probably make the first warning log a debug level log and show a final error after failing permanently. Either by no-retry or by timing out in retry (log last error then).

File: notify/impl.go (hunk 1 above)
Comment (brancz) [inline notify/impl.go]: sgtm

**Comment by beorn7** [comment]
This is on my review queue but I'm currently a bit overwhelmed with my day job. No ETA. Feel free to not wait for me.

**Comment by fabxc** [comment]
All comments full sentences (upper-case and final period). LGTM after comments.

File: notify/impl.go (hunk 4 above)
Comment (brancz) [inline notify/impl.go]: it doesn't make a difference in the implementation, but yes that makes sense

File: notify/impl.go [hunk 6]
```diff
@@ -546,26 +571,33 @@ func (n *Hipchat) Notify(ctx context.Context, as ...*types.Alert) error {
...
 	defer resp.Body.Close()
 
-	if resp.StatusCode/100 != 2 {
-		return fmt.Errorf("unexpected status code %v", resp.StatusCode)
+	return n.retry(resp.StatusCode)
+}
+
+func (n *Hipchat) retry(statusCode int) (bool, error) {
+	// Response code 429 for rate limiting and 5xx can potentially recover.
+	// 2xx responce codes are indicate successful requests.
```
Comment (brancz) [inline notify/impl.go]: are indicate ..

File: notify/impl.go [hunk 7]
```diff
@@ -651,40 +683,45 @@ func (n *OpsGenie) Notify(ctx context.Context, as ...*types.Alert) error {
...
 
 	resp, err := ctxhttp.Post(ctx, http.DefaultClient, apiURL, contentTypeJSON, &buf)
 	if err != nil {
-		return err
+		return true, err
 	}
 	defer resp.Body.Close()
 
-	if resp.StatusCode == 400 && alerts.Status() == model.AlertResolved {
+	// missing documentation therefore assuming 5xx is recoverable
```
Comment (brancz) [inline notify/impl.go]: full sentence & upcase

File: notify/impl.go [hunk 8]
```diff
@@ -227,38 +235,38 @@ func (n *Email) auth(mechs string) (smtp.Auth, error) {
...
 
 	if n.conf.RequireTLS {
 		if ok, _ := c.Extension("STARTTLS"); !ok {
-			return fmt.Errorf("require_tls: true (default), but %q does not advertise the STARTTLS extension", n.conf.Smarthost)
+			return false, fmt.Errorf("require_tls: true (default), but %q does not advertise the STARTTLS extension", n.conf.Smarthost)
 		}
 		tlsConf := &tls.Config{ServerName: host}
 		if err := c.StartTLS(tlsConf); err != nil {
-			return fmt.Errorf("starttls failed: %s", err)
+			return false, fmt.Errorf("starttls failed: %s", err)
```
Comment (fabxc) [inline notify/impl.go]: Mh, this is executing a command. Isn't that generally a request/response cycle and hence can always error due to a flaky network. I think that's recoverable at least in some cases.

Comment (fabxc) [inline notify/impl.go]: Mh, so are the `c.Extension()` calls I guess.

Comment (brancz) [inline notify/impl.go]: True, then I'll make all request/response calls retry. I missed that and only thought of "is a TLS likely to be recovered on a retry?".

Comment (fabxc) [inline notify/impl.go]: Well, not super likely but might just have been a network blip. Exponential backoff is still a relatively safe mechanism and if we fail at this stage, we are at least not spamming any humans.

Comment (brancz) [inline notify/impl.go]: as far as I understand the extension calls actually don't do a request/response (see [here](https://en.wikipedia.org/wiki/Extended_SMTP#Extensions) as a list of supported extensions is sent after a successful EHLO

Comment (brancz) [inline notify/impl.go]: Nevermind, `Extension` [actually calls `hello`](https://golang.org/src/net/smtp/smtp.go?s=10039:10092#L347) which in turn makes it a request/response game again. I propose to make all these retry then. Wdyt?

Comment (fabxc) [inline notify/impl.go]: Yep, let's err on the side of being noisy until proven otherwise.

**Comment by fabxc** [comment]
👍 🚢

**Pull request merged**